mlrun/runtimes/base.py

```diff
@@ -1284,6 +1288,95 @@ def monitor_runs(self, db: DBInterface, db_session: Session):
+        previous run reaches a terminal state
+        """
+        if not run:
+            try:
+                run = db.read_run(db_session, run_uid, project)
+            except mlrun.errors.MLRunNotFoundError:
+                logger.warning(
+                    "Run not found, got deleted by another process",
+                    project=project,
+                    uid=run_uid,
+                )
+                return
+        now = now_date()
+        db_run_state = run.get("status", {}).get("state")
+        if not db_run_state:
+            # can't do any verification due to no run_state updated
+            return
+        if db_run_state not in RunStates.terminal_states():
+            if (
+                resolved_runtime_resources
+                and run_uid in resolved_runtime_resources.get(project, {})
+            ):
+                # if found resource there is no need to continue
+                return
+            last_update_str = run.get("status", {}).get("last_update")
+            debounce_period = (
+                config.resolve_debouncing_period_for_non_terminal_state_run_without_resource()
+            )
+            if last_update_str is not None:
+                last_update = datetime.fromisoformat(last_update_str)
+                run_state = RunStates.gone
+            else:
+                last_update = now
+                run_state = db_run_state
+                logger.info(
+                    "Runs monitoring found run in non-terminal state without last update time set, "
+                    "updating last update time to now, to be able to evaluate next time if something changed",
+                    project=project,
+                    uid=run_uid,
+                    db_run_state=db_run_state,
+                    now=now,
+                    debounce_period=debounce_period,
+                )
+            if last_update > now - timedelta(seconds=debounce_period):
+                # we are setting non-terminal states to runs before the run is actually applied to k8s, that is
+                # why we want to give a grace period in case that is the situation and the resource hasn't been applied
+                logger.warning(
+                    "Monitoring did not discover a runtime resource that corresponded to a run in a "
+                    "non-terminal state. but record has recently updated. Debouncing",
+                    project=project,
+                    uid=run_uid,
+                    db_run_state=db_run_state,
+                    last_update=last_update,
+                    now=now,
+                    debounce_period=debounce_period,
+                )
+            else:
+                logger.info("Updating run state", run_uid=run_uid, run_state=run_state)
+                run.setdefault("status", {})["state"] = run_state
+                run.setdefault("status", {})["last_update"] = now
+                db.store_run(db_session, run, run_uid, project)
```

Comment:
I understand you try to prevent duplicating these two lines:
```
run.setdefault("status", {})["last_update"] = now
db.store_run(db_session, run, run_uid, project)
```
But that is making the code here hard to read IMO, and more important, you're mistakenly logging that you're updating the run state when you're only setting the `last_update`
I suggest to duplicate these lines and have the code better organized 
In addition you should the `last_update` to `now.isoformat()`